### PR TITLE
Test profile input validation (EXPOSUREAPP-8524)

### PR DESCRIPTION
**Pull request by marcauberer**
Adds validation / input constraints to the fields `phone number`, `email`, `date of birth` and `postal code`

- Phone number gets validated with the `PHONE` validator pattern (valid phone number required to save profile)
- Email gets validated with the `EMAIL_ADDRESS` validator pattern (valid email address required to save profile)
- Postal code gets limited to 5 chars as we do not have longer postal codes in Germany
- Date of birth now rejects data selections in the future

Furthermore, this pr adds formatting of the phone number text field for more intuitivity.

We had a discussion about input validation at the QM department last week and the status quo was, that input validation is not planned to implement from SAP side. I thought to make an open source contribution, to provide a solution in case the interest exists ...

*Warning: Up to now the error messages are hard-coded. Those would have to be extracted to the string resources.*

---
Internal Tracking ID: [EXPOSUREAPP-8524](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8524)

### CODE REVIEW

**Comment by harambasicluka** [comment]
Please target `release/2.7.x` as we need the german texts and also the mockups need to be adapted.

**Review by d4rken — DISMISSED** [review]
lgtm
Note that if the user has already entered invalid information, it is still displayed in the edit screen until save is pressed.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/coronatest/rat/profile/create/RATProfileCreateFragment.kt [hunk 1]
```diff
@@ -55,10 +59,41 @@ class RATProfileCreateFragment : Fragment(R.layout.rat_profile_create_fragment),
...
+                if (Patterns.PHONE.matcher(it.toString()).matches()) {
+                    viewModel.phoneChanged(it.toString())
+                } else {
+                    viewModel.phoneChanged("")
+                }
+            }
+            phoneInputEdit.setOnFocusChangeListener { _, hasFocus ->
+                // Validate phone number
+                if (!hasFocus && !Patterns.PHONE.matcher(phoneInputEdit.text.toString()).matches()) {
+                    phoneInputLayout.error = "Invalid phone number"
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/coronatest/rat/profile/create/RATProfileCreateFragment.kt:73]: @UA we would need a german text for this

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/coronatest/rat/profile/create/RATProfileCreateFragment.kt [hunk 2]
```diff
@@ -55,10 +59,41 @@ class RATProfileCreateFragment : Fragment(R.layout.rat_profile_create_fragment),
...
+                if (Patterns.EMAIL_ADDRESS.matcher(it.toString()).matches()) {
+                    viewModel.emailChanged(it.toString())
+                } else {
+                    viewModel.phoneChanged("")
+                }
+            }
+            emailInputEdit.setOnFocusChangeListener { _, hasFocus ->
+                // Validate email
+                if (!hasFocus && !Patterns.EMAIL_ADDRESS.matcher(emailInputEdit.text.toString()).matches()) {
+                    emailInputLayout.error = "Invalid email address"
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/coronatest/rat/profile/create/RATProfileCreateFragment.kt:92]: @UA we would need a german text for this

**Review by NataliaLemmerth — DISMISSED** [review]
UA approved. No text changes.

**Comment by harambasicluka** [comment]
Followed by: https://github.com/corona-warn-app/cwa-app-android/pull/3771

**Pull request closed without merging**